### PR TITLE
context-aware/flexible few-shot example rendering

### PR DESCRIPTION
**Pull request by craterink**
Sometimes you want to inject as many few-shot examples that fit into model context. This idea PR calculates token constraints and renders the maximum number of examples to fit under the model limit.

### CODE REVIEW

File: examples/format_contextaware_example.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                content="Text: I love this product. It's the best thing ever!",
+            ),
+        ],
+        format_variables={
+            "examples": examples,
+        },
+        contextaware_options={
+            "target": "examples",
+            "elem_render": render_example,
+            "elem_separator": "\n\n",
```
Comment (craterink) [inline examples/format_contextaware_example.py:69]: target variable must be an iterator. note this includes strings, so you could implement a chunking summarizing feature with this same setup (though we would need to return/store the number of target elements consumed by the render)

File: gpt_json/gpt.py [hunk 2]
```diff
@@ -137,18 +152,93 @@ async def run(
             to decide whether they want to allow the modifications or not.
 
         """
-        messages = [
-            self.fill_message_template(message, format_variables or {})
-            for message in messages
-        ]
+        if contextaware_options is not None:
```
Comment (craterink) [inline gpt_json/gpt.py:155]: here is the change

**Comment by piercefreeman** [comment]
@craterink I had a similar idea a few days ago and sketched out a stub ticket here: https://github.com/piercefreeman/gpt-json/issues/23. In some ways this case is a subclass of that case: calculating allowable length of the message and then truncating the example payload to the amount of the examples that can fit fully.

The one additional component here might be allowing users to truncate a list of objects (like examples) to only include full list values. Want to append that ticket with this use-case?

**Comment by rdiehlmartinez** [comment]
@piercefreeman second this suggestion. Thanks @craterink keep me in the loop

**Comment by craterink** [comment]
closing bc #28 is simpler version of this

**Pull request closed without merging**